Controls

arora767287 opened the pull request:
Hello,

Can you review the code in this branch and let me know what I should change or if I should start over? The documentation has a diagram that explains what each of the files do and how they interact with each other if it is confusing. I will put it here as well just for reference when you look through the code.

Thanks
![Screen Shot 2021-03-20 at 9 41 58 PM](https://user-images.githubusercontent.com/57432791/111871486-4c80bc80-89c5-11eb-97dc-6b3f34633106.png)

priime0 committed:
fix: add software dummy code

Signed-off-by: Lucas Sta Maria <lucas.stamaria@gmail.com>

priime0 committed:
Merge branch 'deadline-day' of https://github.com/One-Degree-North/Team-E into deadline-day

arora767287 committed:
Create controls.py

arora767287 committed:
Delete controls_2.py

arora767287 committed:
Update controls.py

arora767287 committed:
Update object.py

arora767287 committed:
Create camera.py

arora767287 committed:
Create interface.py

arora767287 committed:
Rename main.py to temporary_main.py

arora767287 committed:
Create main.py

arora767287 committed:
Update main.py

arora767287 committed:
Update README.md

arora767287 committed:
Update README.md

arora767287 committed:
Update camera.py

arora767287 committed:
Update controls.py

arora767287 committed:
Update interface.py

arora767287 committed:
Update main.py

arora767287 committed:
Update object.py

priime0 commented on main.py:
```diff
@@ -0,0 +1,48 @@
+        camera.release()
+    
+    
+'''
```
Consider removing this `'''` as it will throw an error.

priime0 commented on main.py:
```diff
@@ -0,0 +1,48 @@
+    ))
+
+if __name__ == '__main__':
+    camera = Camera()
```
An import needs to be made for `Camera()`.

priime0 commented on main.py:
```diff
@@ -0,0 +1,48 @@
+    camera = Camera()
+    camera.open(gstreamer_pipeline(sensor_id=0, sensor_mode=3, flip_method=0, display_height=600, display_width=600))
+    camera.start()
+    interface = Interface()
```
`Interface` also needs to be imported.

priime0 replied:
Also missing arguments to instantiate the `Interface` object.

priime0 commented on main.py:
```diff
@@ -0,0 +1,48 @@
+    controller = control(100,0.1)
+    interface.make_interface() 
+    try:
+        controls_move()
```
Should be `controller.controls_move()`.

priime0 commented on main.py:
```diff
@@ -0,0 +1,48 @@
+    except:
+        print("Unexpected error: ", sys.exc_infor()[0])
+        raise
+    if(not camera.video_capture.isOpened()):
```
Parenthesis enclosing `not camera.video_capture.isOpened()` may be unnecessary here.

priime0 commented on object.py:
```diff
@@ -10,9 +10,6 @@
   lines_video = picture_lines(picture_video)
   region_place_part = region_place(lines, lines_video)
   draw_picture("/Users/nityaarora/Downloads/picture.png", region_place_part[0], region_place_part[1], region_place_part[2], region_place_part[3])
```
Might want to use relative pathing instead of absolute pathing (replace `/Users/nityaarora/` with `~/` so that other people can run the code.

priime0 commented on object.py:
```diff
@@ -10,9 +10,6 @@
-
-  
-  
 def picture_lines(pictures):
```
Maybe add spacing above the function for spacing and readability.

priime0 commented on controls.py:
```diff
@@ -0,0 +1,110 @@
+        
+        self.clock = pygame.time.Clock()
+        
+        self.serial_port = "dev/ttyUSB0"
```
I think this should be "`/dev/ttyUSB0`"

priime0 commented on controls.py:
```diff
@@ -0,0 +1,110 @@
+        
+        self.serial_baudrate = 9600
+        
+    ser = serial.Serial(port=serial_port, baudrate=serial_baudrate)    
```
Indent this line and consider using `self.ser`.

priime0 commented on controls.py:
```diff
@@ -0,0 +1,110 @@
+            elif raw < 0:
+                return  (raw+self.adjustment)/1.0       
+                
+    #Creates variables that store the motor labels for the different motors used. Some motor values have also been put in for adjustments that may be made to add the claw          
```
If the below variables are constants, consider using `SCREAMING_SNAKE_CASE` (style recommendation). Also consider putting these variables closer to the top of the module and prepending their names with `_`.

priime0 commented on controls.py:
```diff
@@ -0,0 +1,110 @@
+    left_motor = 3
+    up_quantity = 500
+    right_quantity = 500
+    left_quantity = 500
```
`left_quantity` is never used.

priime0 commented on controls.py:
```diff
@@ -0,0 +1,110 @@
+    central_motor = 1
+    right_motor = 2
+    left_motor = 3
+    up_quantity = 500
```
`up_quantity` is never used.

priime0 commented on controls.py:
```diff
@@ -0,0 +1,110 @@
+                #The axis labelled 4 is used for up to right motion and is located on the right joystick.
+                elif event.type == pygame.JOYAXISMOTION:
+                    input = event.value
+                    scaled_input = movement_scaler(input, 0.2)
```
You can condense lines `89` and `90`. There doesn't appear to be a difference in value between `right_quantity` and `scaled_input` after line 89.

priime0 commented on controls.py:
```diff
@@ -0,0 +1,110 @@
+class control():
```
Style recommendation: consider using `UpperCamelCase` for class names.

Also, the `control` class is never used in any other file. Need to import it somewhere.

priime0 commented on interface.py:
```diff
@@ -0,0 +1,25 @@
+class interface():
```
Style recommendation: consider using `UpperCamelCase` for class names.

`main.py` uses `Interface()` (which it also needs to import).

priime0 commented on interface.py:
```diff
@@ -0,0 +1,25 @@
+        self.running = True
+        self.cameraFrameSize = cameraFrameSize
+        self.graphFrameSize = graphFrameSize
+        while self.running:
```
When will `self.running` become `False`?

Also, this infinite loop might need a thread allocated to it, so the rest of the program can run.

arora767287 replied:
self.running is false when the window is closed (when WIN_CLOSED is detected)

arora767287 replied:
I think I fixed the second comment

priime0 commented on camera.py:
```diff
@@ -0,0 +1,56 @@
+class Camera:
```
Import this in `main.py`.

Also add spacing between the functions for better readability.

priime0 commented on camera.py:
```diff
@@ -0,0 +1,56 @@
+    #This opens up the camera stream to allow frames to be analyzed through gstreamer
+    def open(self, gstreamer_pipeline_string):
+        try:
+            self.video_capture = cv.VideoCapture(gstreamer_pipeline_string, cv2.CAP_GSTREAMER)
```
`cv2` isn't imported.

priime0 replied:
Also, is there a difference between `cv` and `cv2`?

arora767287 replied:
Just a version difference, cv2 has more functionality I think. I've never actually used cv. I think it is a very old version.

priime0 commented on camera.py:
```diff
@@ -0,0 +1,56 @@
+            self.running = True
+            self.read_thread = threading.Thread(target = self.updateCamera)
+            self.read_thread.start()
+        return self  
```
Is there a need to `return self`?

arora767287 replied:
To use it for chaining later on.

priime0 commented on camera.py:
```diff
@@ -0,0 +1,56 @@
+            return frame, grabbed
+    #This ends the camera stream and lets the camera stop outputting frames by setting the video capture object to none        
+     def release(self):
+         if self.video_capture!=None:
```
Style recommendation: add spaces here for readability.

priime0 commented on temporary_main.py:
```diff
@@ -0,0 +1,82 @@
+import pygame
```
You can delete this file and access it later with Git.

priime0 commented on camera.py:
```diff
@@ -0,0 +1,63 @@
+            grabbed = self.grabbed
+            return frame, grabbed
+    #This ends the camera stream and lets the camera stop outputting frames by setting the video capture object to none        
+     def release(self):
```
Incorrect indentation here by the way.

arora767287 committed:
Update camera.py

arora767287 committed:
Update interface.py

arora767287 committed:
Update interface.py

arora767287 committed:
Update controls.py

arora767287 committed:
Update object.py

arora767287 committed:
Update main.py

arora767287 committed:
Update main.py

arora767287 committed:
Update main.py

arora767287 committed:
Delete temporary_main.py

arora767287 committed:
Update main.py

arora767287 committed:
Update main.py

arora767287 committed:
Update main.py

arora767287 committed:
Update interface.py

arora767287 committed:
Update main.py

arora767287 committed:
Update main.py

arora767287 committed:
Create main.py

arora767287 committed:
Update interface.py

arora767287 committed:
Update interface.py

arora767287 committed:
Update main.py

arora767287 committed:
Update main.py

arora767287 committed:
Update controls.py

priime0 commented on camera.py:
```diff
@@ -0,0 +1,69 @@
+import numpy
+import matplotlib
+import PySimpleGUI as sg
+import time
```
Add spacing between imports and class.

priime0 commented on camera.py:
```diff
@@ -0,0 +1,69 @@
+    #This opens up the camera stream to allow frames to be analyzed through gstreamer
+    def open(self, gstreamer_pipeline_string):
+        try:
+            self.video_capture = cv2.VideoCapture(gstreamer_pipeline_string, cv2.CAP_GSTREAMER)
```
Still missing a `cv2` import.

priime0 commented on controls.py:
```diff
@@ -0,0 +1,118 @@
+import numpy
+import matplotlib
+import PySimpleGUI as sg
+import time
```
Add spacing between imports and class.

priime0 commented on controls.py:
```diff
@@ -0,0 +1,118 @@
+        
+        self.clock = pygame.time.Clock()
+        
+        self.serial_port = "/dev/ttyUSB0"
```
Should this be `self.serial_port = serial_port`? Same with with `self.serial_baudrate`.

arora767287 replied:
Yeah it should

priime0 commented on interface.py:
```diff
@@ -0,0 +1,41 @@
+        self.camera_column = sg.Column(camera_layout, element_justification="center")
+        self.speed_graph_layout = [[sg.Text("Speeds Graphs", size=(60,60), justification="center")], [sg.Image(filename="graph.png", key="graphs")]]
+        self.layout = [camera_layout, speed_graph_layout]
+        self.window = sg.Window(title="Interface", layout=layout, return_keyboard_events=True,location=(600,600))
```
Add spacing between functions.

priime0 commented on interface.py:
```diff
@@ -0,0 +1,41 @@
+    #make_interface makes the user interface that will use information parsed through from the camera 
+    #and the controls in order to make information about the robot's speed, acceleration, and other measurements available
+    def make_interface(self):
+        
```
Remove this line here.

priime0 commented on interface.py:
```diff
@@ -0,0 +1,41 @@
+    
+    def __init__(self, robot, object_seen, motor_controls):
+        
+        self.robot = None
```
Should this be `self.robot = robot`, etc.?

arora767287 replied:
Yeah

priime0 commented on main.py:
```diff
@@ -0,0 +1,63 @@
+import time
+from cameras.py import Camera
+from interface.py import Interface
+from controls.py import Control
```
Add space between imports and rest of the file.

priime0 commented on main.py:
```diff
@@ -0,0 +1,63 @@
+        
+        camera.stop()
+        camera.release()
+    
```
Remove extra spacing at the bottom.

priime0 commented on main.py:
```diff
@@ -0,0 +1,63 @@
+import PySimpleGUI as sg
+from threading import Thread
+import time
+from cameras.py import Camera
```
Remove the `.py` extensions in these lines. Also it's `camera` not `cameras`.

arora767287 committed:
Update main.py

arora767287 committed:
Update interface.py

arora767287 committed:
Update controls.py

arora767287 committed:
Update controls.py

arora767287 committed:
Update camera.py

priime0 commented on controls.py:
```diff
@@ -0,0 +1,119 @@
+    _RIGHT_MOTOR = 2
+    _LEFT_MOTOR = 3
+    _UP_QUANTITY = 500
+    _RIGHT_QUANTITY = 500
```
`_RIGHT_QUANTITY` isn't a constant. Maybe this should be `self.right_quantity` if you want it to be an instance's variable? Else you can just remove it here and make the `_RIGHT_QUANTITY` below be `right_quantity`, since its scope won't be global. 

`_UP_QUANTITY` and `_LEFT_QUANTITY` are still unused. Consider removing.

priime0 commented on controls.py:
```diff
@@ -0,0 +1,119 @@
+                #The axis labelled 4 is used for up to right motion and is located on the right joystick.
+                elif event.type == pygame.JOYAXISMOTION:
+                    input = event.value
+                    _RIGHT_QUANTITY = movement_scaler(input, 0.2)
```
If you remove `_RIGHT_QUANTITY` above and rename this to `right_quantity`, its scope will only be in the function.

priime0 commented on main.py:
```diff
@@ -0,0 +1,61 @@
+    while self.running:
+        _, camera_picture = camera.read()
+        camera_pictures = np.hstack((camera_picture))
+        thread = Thread(target=interface.update_interface(cameraFrameSize=(600,600), videoFrames=camera_pictures))
```
Move this outside the `while` loop; you're declaring a new thread with each iteration. Also remember to call `thread.join()` at the end of the program.

priime0 commented on main.py:
```diff
@@ -0,0 +1,61 @@
+    camera = Camera()
+    camera.open(gstreamer_pipeline(sensor_id=0, sensor_mode=3, flip_method=0, display_height=600, display_width=600))
+    camera.start()
+    interface = Interface(robot=camera.list_movements, object_seen=True, motor_controls=True)
```
`camera` doesn't have a `list_movements` field.

arora767287 committed:
Added button variables

I started editing for the claw controls and then realized that the team hasn't decided on how the claw will be controlled (based on the documentation)

arora767287 committed:
Fixed threading mistake

arora767287 committed:
Create functions.py

priime0 commented on main.py:
```diff
@@ -0,0 +1,61 @@
+    while self.running:
+        _, camera_picture = camera.read()
+        camera_pictures = np.hstack((camera_picture))
+        thread = Thread(target=update_interface, args=(cameraFrameSize=(600,600), videoFrames=camera_pictures,))
```
`update_interface` doesn't exist.

arora767287 replied:
Now I see them

arora767287 committed:
Update interface.py

arora767287 committed:
Update controls.py

arora767287 committed:
Update main.py

arora767287 committed:
Thread arguments

arora767287 committed:
Update functions.py

arora767287 committed:
Update functions.py

arora767287 committed:
Update interface.py

arora767287 committed:
Update interface.py

arora767287 committed:
Update main.py

priime0 commented on interface.py:
```diff
@@ -0,0 +1,139 @@
+        global power
+        global t
+        
+        angular_position = angular_positon.append(self.information.angular_position_changer(angular_position_arrays[1])))
```
There's one too many closing parentheses here.

priime0 commented on functions.py:
```diff
@@ -0,0 +1,41 @@
+    self.clock = clock
+    self.start_time = clock.time.time() #There will be a little bit of lag since this function will be called after interface is called
+   
+  def speed_estimates(self, C_d, density, frontal_area, andgle_of_attack):
```
Type here: should be `angle_of_attack`.

priime0 commented on functions.py:
```diff
@@ -0,0 +1,41 @@
+  def speed_estimates(self, C_d, density, frontal_area, andgle_of_attack):
+    self.speed_left = math.sqrt((2*self.robot["left_motor"])/(C_d*density*frontal_area*sin(angle_of_attack)))
+    self.speed_right = math.sqrt((2*self.robot["right_motor"])/(C_d*density*frontal_area*sin(angle_of_attack)))
+    self.speed_up = math.sqrt((2*self.robot["up_motor"])/(C_d*density*frontal_area*sin(angle_of_attack))
```
Missing closing parenthesis here.

priime0 commented on main.py:
```diff
@@ -0,0 +1,63 @@
+    camera = Camera()
+    camera.open(gstreamer_pipeline(sensor_id=0, sensor_mode=3, flip_method=0, display_height=600, display_width=600))
+    camera.start()
+    controller = Control(100,0.1)
```
`Control` is missing many arguments.

arora767287 committed:
Update interface.py

arora767287 committed:
Update functions.py

arora767287 committed:
Update functions.py

arora767287 committed:
Update controls.py

priime0 commented on controls.py:
```diff
@@ -0,0 +1,118 @@
+        self.adjustment = adjustment
+        self.running = False
+        self.list_movements = {
+            abutton: 0,
```
Each key in the dictionary should be enclosed in `'`.

priime0 commented on controls.py:
```diff
@@ -0,0 +1,118 @@
+        }
+        
+        #Initializes the joystick being used, which is the XBox Controller
+        
```
`pygame.joystick.init()` should be called here to initialize the joystick system.

priime0 commented on controls.py:
```diff
@@ -0,0 +1,118 @@
+        
+        self.controller = pygame.joystick.Joystick(0)
+        #Sets up the timing for the interface
+        self.A = controller.get_button(0)
```
Should be `self.controller`, not just `controller`.

priime0 commented on controls.py:
```diff
@@ -0,0 +1,118 @@
+        
+        #Initializes the joystick being used, which is the XBox Controller
+        
+        self.controller = pygame.joystick.Joystick(0)
```
It might not be a good idea to hard-code the controller here, as `pygame` might throw an error if the controller is not connected.

What you could instead do is have a list of `controllers` with `pygame.joystick.get_count()` (see Pygame documentation). Pygame events such as `pygame.CONTROLLERDEVICEADDED`/`pygame.JOYDEVICEADDED` and `pygame.CONTROLLERDEVICEREMOVED`/`pygame.JOYDEVICEREMOVED` can also be used.

arora767287 replied:
Does using the if statement work, or do I have to use the events?

priime0 commented on functions.py:
```diff
@@ -0,0 +1,41 @@
+    self.start_time = clock.time.time() #There will be a little bit of lag since this function will be called after interface is called
+   
+  def speed_estimates(self, C_d, density, frontal_area, angle_of_attack):
+    self.speed_left = math.sqrt((2*self.robot["left_motor"])/(C_d*density*frontal_area*sin(angle_of_attack)))
```
Add spacing for readability and consider using a function to perform the calculations for readability as well.

priime0 commented on functions.py:
```diff
@@ -0,0 +1,41 @@
+    return [tdata, velocity_found]
+    
+  def angular_position_changer(time, angular_velocity):
+  
```
Remove this line.

priime0 commented on functions.py:
```diff
@@ -0,0 +1,41 @@
+  
+  
+  def angular_position_arrays(self):
+  
```
Remove this line.

priime0 commented on functions.py:
```diff
@@ -0,0 +1,41 @@
+    self.move_angle = self.robot["move_turn"]
+    time_new = self.clock.time.time()
+    velocity_data = []
+    while len(velocity_data)<6 and time_new-time_start < 1:
```
Spacing inconsistent.

priime0 commented on interface.py:
```diff
@@ -0,0 +1,139 @@
+        #
+        t=zeros(0)
+
+        # set plots
```
Reduce spacing between lines here.

arora767287 committed:
Update controls.py

arora767287 committed:
Update functions.py

arora767287 committed:
Update interface.py

arora767287 committed:
Update controls.py

priime0 commented on controls.py:
```diff
@@ -0,0 +1,130 @@
+                # get_instance_id() is an SDL2 method
+                jid = joystick.get_id()
+                return AttributeError
+         self.controller = joysticks[0]                              
```
Fix indentation.

priime0 replied:
This will still throw an error if a joystick isn't connected.

arora767287 committed:
Update controls.py

priime0 commented on main.py:
```diff
@@ -0,0 +1,63 @@
+    camera.open(gstreamer_pipeline(sensor_id=0, sensor_mode=3, flip_method=0, display_height=600, display_width=600))
+    camera.start()
+    controller = Control(100,0.1)
+    interface = Interface(robot=controller.list_movements, object_seen=True, motor_controls=True)
```
Missing an argument here.

priime0 commented on interface.py:
```diff
@@ -0,0 +1,137 @@
+    
+    def __init__(self, robot, object_seen, motor_controls, clock):
+        self.clock = clock
+        self.graphs_figure = figure(num = 0, figsize = (12, 8))
```
You need to import or declare `figure`.

priime0 commented on interface.py:
```diff
@@ -0,0 +1,137 @@
+        self.robot = robot
+        self.object_seen = object_seen
+        self.motor_controls = motor_controls
+        self.graph_angular_position = plt.figure()
```
You need to import `plt`.

arora767287 committed:
Update main.py

arora767287 committed:
Update interface.py

arora767287 committed:
Create interface.py

priime0 commented on interface.py:
```diff
@@ -0,0 +1,138 @@
+import time
+from threading import Thread
+import matplotlib.animation as animation
+from plt import figure
```
`plt` also needs to be imported.

priime0 replied:
This needs to be `matplotlib.pyplot`, doesn't it?

priime0 commented on main.py:
```diff
@@ -0,0 +1,63 @@
+    camera.open(gstreamer_pipeline(sensor_id=0, sensor_mode=3, flip_method=0, display_height=600, display_width=600))
+    camera.start()
+    controller = Control(100,0.1)
+    interface = Interface(robot=controller.list_movements, object_seen=True, motor_controls=True, clock=controller.clock)
```
If the controller isn't connected, passing `controller.clock` will fail because `clock` was never instantiated. Check the `Control` class.

arora767287 committed:
Update interface.py

arora767287 committed:
Update main.py

arora767287 committed:
Update controls.py

arora767287 committed:
Update interface.py

priime0 commented on interface.py:
```diff
@@ -0,0 +1,139 @@
+    
+    def __init__(self, robot, object_seen, motor_controls, clock):
+        self.clock = clock
+        self.graphs_figure = figure(num = 0, figsize = (12, 8))
```
`plt.figure`

priime0 commented on interface.py:
```diff
@@ -0,0 +1,139 @@
+        self.graph_angular_velocity = plt.figure()
+        self.graph_acceleration = plt.figure() 
+        self.graph_power = plt.figure()
+        self.graphs = [self.angular_position, self.graph_speed, self.graph_position, self.graph_angular_velocity, self.graph_acceleration, self.graph_power]
```
There is no `angular_position` attribute.

priime0 commented on functions.py:
```diff
@@ -0,0 +1,39 @@
+import matplotlib
+import matplotlib.pyplot as plt
+import numpy as np
+from scipy.optimize import curve_fit
```
Add spacing below this.

priime0 commented on functions.py:
```diff
@@ -0,0 +1,39 @@
+import matplotlib.pyplot as plt
+import numpy as np
+from scipy.optimize import curve_fit
+class interfaceInformation():
```
Use `UpperCamelCase`.

priime0 commented on functions.py:
```diff
@@ -0,0 +1,39 @@
+  #Initialize with the clock system of the interface and robot
+  def __init__(self, clock): 
+    self.clock = clock
+    self.start_time = clock.time.time() #There will be a little bit of lag since this function will be called after interface is called
```
Did you mean `get_time()`? `time` does not exist.

priime0 commented on interface.py:
```diff
@@ -0,0 +1,139 @@
+    #and the controls in order to make information about the robot's speed, acceleration, and other measurements available
+    def make_interface(self):
+        self.camera_layout = [[sg.Text(title="Camera", size=(60,60), justification="center")], [sg.Image(filename="", key="camera")]]
+        self.camera_column = sg.Column(camera_layout, element_justification="center")
```
Probably meant to be `self.camera_layout`.

priime0 commented on interface.py:
```diff
@@ -0,0 +1,139 @@
+    #make_interface makes the user interface that will use information parsed through from the camera 
+    #and the controls in order to make information about the robot's speed, acceleration, and other measurements available
+    def make_interface(self):
+        self.camera_layout = [[sg.Text(title="Camera", size=(60,60), justification="center")], [sg.Image(filename="", key="camera")]]
```
Don't need the `title` keyword argument for `sg.Text`.

priime0 commented on interface.py:
```diff
@@ -0,0 +1,139 @@
+        self.camera_layout = [[sg.Text(title="Camera", size=(60,60), justification="center")], [sg.Image(filename="", key="camera")]]
+        self.camera_column = sg.Column(camera_layout, element_justification="center")
+        self.speed_graph_layout = [[sg.Text("Graphs", size=(60,60), justification="center")], [sg.Canvas(key="Canvas")]]
+        self.layout = [camera_layout, speed_graph_layout]
```
Missing `self` here as well.

priime0 commented on main.py:
```diff
@@ -0,0 +1,64 @@
+    except KeyboardInterrupt:
+        print("Stop")
+    except:
+        print("Unexpected error: ", sys.exc_infor()[0])
```
Need to import `sys`.

priime0 commented on controls.py:
```diff
@@ -0,0 +1,133 @@
+    #Starts controls and sets up event responses to control movements on the controller
+    def controls_move(self):
+        self.running = True
+        while running:
```
Add `self`.

priime0 commented on controls.py:
```diff
@@ -0,0 +1,133 @@
+            claw_movement = 99
+            for event in pygame.event.get():
+                if event.type == pygame.QUIT:
+                    running = False
```
Add `self`.

priime0 commented on controls.py:
```diff
@@ -0,0 +1,133 @@
+
+            self.serial_baudrate = serial_baudrate
+
+            self.ser = serial.Serial(port=serial_port, baudrate=serial_baudrate)    
```
Need to open `ser` to use it.

arora767287 committed:
Update functions.py
